dev-server.js

```diff
@@ -6,52 +6,108 @@ const compiler = webpack(conf)
 const express = require('express')
 const app = express()
 const bodyParser = require('body-parser')
-app.use(bodyParser.text({type: '*/*'}))
+const url = require('url')
+app.use(bodyParser.text({ type: '*/*' }))
 
 const { exec } = require('child_process')
 
+/**
+ * Constants.
+ */
+var SERVER_SOFTWARE = 'Node/' + process.version
+var SERVER_PROTOCOL = 'HTTP/1.1'
+var GATEWAY_INTERFACE = 'CGI/1.1'
+
 const perl = (req, res) => {
+  const file = resolve.apply(null, [__dirname].concat(req.url.split('/')))
+
+  if (!req.hasOwnProperty('uri')) {
+    req.uri = url.parse(req.url)
+  }
+
+  var host = (req.headers.host || '').split(':')
+  var address = host[0]
+  var port = host[1]
 
-  const sendErr = msg => res.status(500).send(msg || "Request Failed")
-  const file = resolve.apply(null, [__dirname].concat(req.url.split("/")))
-  exec(`carton exec ${file} 'POSTDATA=${req.body}'`, {cwd: file.substring(0, file.lastIndexOf("/")), maxBuffer: Infinity}, (err, stdout, stderr) => {
-    try {
-      if (err) {
-        console.error(err)
-        sendErr(err.message)
-      } else if (stderr) {
-        console.error(stderr)
-        sendErr(stderr)
-      } else if (stdout) {
-        const payload = stdout.replace(/^[^\{]*/g, "")
-        payload
-          ? res.json(JSON.parse(payload))
-          : res.status(200).send("OK")
-      } else {
-        res.status(400).send("Request Failed")
+  var env = {}
+
+  // These meta-variables below can be overwritten by a
+  // user's 'env' object in options
+  Object.assign(env, {
+    GATEWAY_INTERFACE: GATEWAY_INTERFACE,
+    SCRIPT_NAME: file,
+    SERVER_NAME: address || 'unknown',
+    SERVER_PORT: port || 80,
+    SERVER_PROTOCOL: SERVER_PROTOCOL,
+    SERVER_SOFTWARE: SERVER_SOFTWARE,
+  })
+
+  // The client HTTP request headers are attached to the env as well,
+  // in the format: "User-Agent" -> "HTTP_USER_AGENT"
+  for (var header in req.headers) {
+    var name = 'HTTP_' + header.toUpperCase().replace(/-/g, '_')
+    env[name] = req.headers[header]
+  }
+
+  // These final environment variables take precedence over user-specified ones.
+  env.REQUEST_METHOD = req.method.toUpperCase()
+  env.QUERY_STRING = req.uri.query || ''
+  env.REMOTE_ADDR = req.connection.remoteAddress
+  env.REMOTE_PORT = req.connection.remotePort
+  if ('content-length' in req.headers) {
+    env.CONTENT_LENGTH = req.headers['content-length']
+  }
+  if ('content-type' in req.headers) {
+    env.CONTENT_TYPE = req.headers['content-type']
+  }
+  if ('authorization' in req.headers) {
+    var auth = req.headers.authorization.split(' ')
+    env.AUTH_TYPE = auth[0]
+  }
+
+  const sendErr = msg => res.status(500).send(msg || 'Request Failed')
+  exec(
+    `carton exec ${file} 'POSTDATA=${req.body}'`,
+    { cwd: file.substring(0, file.lastIndexOf('/')), maxBuffer: Infinity, env },
```

Comment:
Right, the `env` there tucked sneakily at the end of line 71 should have all of the environment variables in there. I'd prefer not to hardcode them if possible and use the headers from the request in to set them. Seems more future-proof this way.

Reply:
That is a great idea, can we read the headers into a Javascript object?  We can then just stick the resulting variable into env.  Is there a headerParse like bodyParse?

Reply:
Oh and shouldn't that be `env: env`?  Not plain env as you have it.

Reply:
And sorry for being dense, I see you are reading the headers.  Perhaps `env` instead of `env: env` is the problem?

Reply:
The simple `env` should be correct as long as `env` is in the scope. (this changed with ES6 of JS)
E.g., the following is valid:
```javascript
var name = 'test';
var obj = {name} 
console.log(obj.name); // "test"
```
Since I declared the env above, I don't think this is the problem. I'm trying to stringify them now as a series of `exports` and run that before executing the carton command. Maybe the interplay of carton and node's exec drops the environment?? Still unclear.

Will push my progress and then need to take a brake for a bit.

Reply:
Ha. I just realized a typed "brake" instead of "break". I really _do_ need a break I guess!